components/raftstore/src/store/peer.rs

```diff
@@ -3344,6 +3332,87 @@ where
         assert!(self.pending_merge_state.is_some());
         self.want_rollback_merge_peers.insert(peer_id);
     }
+
+    // Check disk usages for the peer itself and other peers in the raft group.
+    // The return value indicates whether the propose is allowed or not.
+    fn check_disk_usages_before_propose_normal<T>(
+        &mut self,
+        ctx: &mut PollContext<EK, ER, T>,
+        disk_full_opt: DiskFullOpt,
+        disk_full_stores: &mut Vec<u64>,
+    ) -> bool {
+        // Propose can be rejected if of the store's disk is full.
+        let self_allowed = match disk_full_opt {
+            DiskFullOpt::NotAllowedOnFull => matches!(ctx.self_disk_usage, DiskUsage::Normal),
+            _ => !matches!(ctx.self_disk_usage, DiskUsage::AlreadyFull),
+        };
+        if !self_allowed {
+            disk_full_stores.push(ctx.store.id);
+        }
+
+        // Then if majority peers for the raft group meets the disk full problem,
+        // reject the propose.
+        let mut check_majority_disk_full = || {
+            let self_peer_id = self.peer_id();
+            let peers_len = self.get_store().region().get_peers().len();
+
+            let mut disk_full_peers = HashSet::default();
+            let mut next_idxs = Vec::with_capacity(peers_len);
+            for peer in self.get_store().region().get_peers() {
+                let (peer_id, store_id) = (peer.get_id(), peer.get_store_id());
```

Comment:
Should only consider voter?

Reply:
It's unnecessary to handle learners specially because we use `has_quorum` to judge a quorum.